### PR TITLE
[Bugfix][Hardware][Intel] Disable compiled grouped_topk on XPU

### PR DESCRIPTION
**Pull request by SergiioB**
## Purpose

On the affected Intel XPU stack, compiling the Python `grouped_topk` router produced invalid expert IDs during NemotronH native-MTP draft warmup:

```text
Assertion `index out of bounds: 0 <= tmp0 < 128` failed.
```

The base WNA16 MoE path can avoid this Python router through the fused XPU implementation, while the MTP draft fallback reaches it during engine compile/warmup.

**Fix:** keep `grouped_topk` eager only on XPU with `torch.compiler.disable`. Other platforms retain the existing `torch.compile` decorator. This is a narrow correctness workaround pending an XPU-safe compiled implementation; it does not clamp IDs or disable compilation for the surrounding model.

The unrelated `test_outputs.py` portability change was removed during rebase because upstream #52237 already contains it.

## Test plan

```bash
pytest -q tests/kernels/moe/test_grouped_topk_xpu.py
```

The XPU-only coverage checks softmax/sigmoid with and without correction bias, validates finite weights and expert IDs in `[0, 128)`, and verifies that an outer compiled graph can cross the XPU-eager router without a device assertion. Non-XPU environments skip this file; existing CUDA coverage is unchanged.

## Reproduction

The reliable trigger is the compiled NemotronH MTP draft warmup path. A standalone random decomposition of the individual top-k operations did not reproduce the assertion, so it is not presented as the reproducer.

Environment and exact command are posted in the PR discussion.

## Validation status

- Branch rebased onto current `main`; PR is mergeable.
- Python syntax compilation and `git diff --check`: pass.
- Focused local pytest was not available in this checkout (`pytest` is not installed).
- Intel CI and upstream checks were retriggered by the rewritten branch.

### CODE REVIEW

**Comment by mayuyuace** [comment]
Can you provide detail CMD to trigger this bug?

**Comment by SergiioB** [comment]
Thanks. The reliable trigger I have is the compiled NemotronH native-MTP draft warmup path; a standalone random decomposition of the individual top-k operations did not reproduce it.

Environment used:

```text
GPU: Intel Arc Pro B70
image: vllm/vllm-openai-xpu@sha256:1da0a95485455f08588c11080b9718992fd7d434c6a965d74654903a9d999c57
vLLM: 0.26.1rc1.dev668+g3ee2df303
vllm-xpu-kernels: 0.1.12.3
model: SergiioB/Nemotron-3.5-Lightning-30B-A3B-GPTQ-INT4-G64-sym
```

Run compiled mode (do **not** pass `--enforce-eager`):

```bash
docker run --rm --network host \
  --device /dev/dri --ipc host --shm-size 16g \
  -v /path/to/Nemotron-3.5-Lightning-30B-A3B-GPTQ-INT4-G64-sym:/model:ro \
  -e ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE \
  -e ZE_AFFINITY_MASK=0 \
  -e SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS=0 \
  -e SYCL_CACHE_PERSISTENT=0 \
  vllm/vllm-openai-xpu@sha256:1da0a95485455f08588c11080b9718992fd7d434c6a965d74654903a9d999c57 \
  vllm serve /model \
    --served-model-name nemotron \
    --dtype bfloat16 \
    --quantization gptq \
    --max-model-len 16384 \
    --max-num-seqs 1 \
    --max-num-batched-tokens 8192 \
    --gpu-memory-utilization 0.90 \
    --no-enable-prefix-caching \
    --speculative-config '{"method":"mtp","num_speculative_tokens":2}' \
    --host 127.0.0.1 --port 8000
```

Pre-fix behavior: EngineCore fails during compile/warmup before the endpoint becomes healthy. The generated XPU Inductor kernel repeatedly asserts:

```text
Assertion `index out of bounds: 0 <= tmp0 < 128` failed.
```

The base WNA16 MoE can load through its fused XPU path; the native MTP draft reaches the Python `grouped_topk` router. This PR leaves that router eager only on XPU while allowing the surrounding model to remain compiled.

I also rebased the PR onto current `main` and removed the unrelated `test_outputs.py` change because #52237 already fixed it upstream.

**Comment by mayuyuace** [comment]
Since XPU currently operates primarily in eager mode, compilation cannot be directly disabled here, as doing so would degrade the performance of other models.
XPU should use a kernel to replace this compiled function, and we have plan to do this in a later version.
The bug encountered in this PR appears to be an issue with `torch.compile`; I will attempt to reproduce it and file an issue with Torch.
Thanks for your work, and please use eager mode to run this model before this bug being fixed.

**Comment by mayuyuace** [comment]
I found that disabling compile cannot fix the error.
The root cause is here: https://github.com/intel/torch-xpu-ops/issues/4839, torch.topk on XPU returns invalid (-1) indices with NaN input at batch size >= ~512.
I have asked torch xpu team to fix it.

**Comment by mganczarenko** [comment]
CC: @astachowiczhabana

**Comment by mayuyuace** [comment]
I found the root cause that why you meet the error:
The SergiioB/Nemotron-3.5-Lightning-30B-A3B-GPTQ-INT4-G64-sym GPTQ-quantized checkpoint lacks the specific weights required for MTP (Multi-Token Prediction)—such as the MTP fusion layers `eh_proj`, `enorm`, and `hnorm`.
Consequently, although vLLM instantiates these layers, the weights remain unpopulated, containing only random, uninitialized data (including NaNs) in memory. This causes the initial MTP forward pass to output all NaNs, which subsequently propagate through the KV cache or SSM state, contaminating all ensuing decoding steps.
Finally, `topk` receives an input containing `NaN` and produces an index of -1, ultimately resulting in an out-of-bounds error.
**Conclusion**:
This GPTQ checkpoint does not contain MTP weights and cannot be used with `--speculative-config mtp`.

**Comment by jikunshang** [comment]
@mayuyuace can we use grouped_topk kernel for xpu now?

**Comment by mayuyuace** [comment]
> @mayuyuace can we use grouped_topk kernel for xpu now?

Latest vllm-xpu-kernels supports grouped_topk kernel.
You need to bump up it firstly.

**Comment by mayuyuace** [comment]
> @mayuyuace can we use grouped_topk kernel for xpu now?

And torch-xpu has fixed the bug of torch.topk.
https://github.com/intel/torch-xpu-ops/pull/4986

File: tests/kernels/moe/test_grouped_topk_xpu.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    """An outer compiled graph can cross the XPU-eager router safely."""
+    from vllm.model_executor.layers.fused_moe.router.grouped_topk_router import (
+        grouped_topk,
+    )
+
+    assert getattr(grouped_topk, "_torchdynamo_disable", False), (
+        "grouped_topk must remain Dynamo-disabled on XPU"
+    )
+
+    torch.manual_seed(1)
```
Comment (AndreasKaratzas) [inline tests/kernels/moe/test_grouped_topk_xpu.py:139]: probably best if you use the test utils seeder function.

File: tests/kernels/moe/test_grouped_topk_xpu.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+from __future__ import annotations
+
+import pytest
+import torch
+
+from vllm.platforms import current_platform
+
+pytestmark = pytest.mark.skipif(
+    not current_platform.is_xpu(),
```
Comment (AndreasKaratzas) [inline tests/kernels/moe/test_grouped_topk_xpu.py:16]: Thank you for that 😅 Some forget to add a file level skip and save CI resources

**Comment by SergiioB** [comment]
Closing this PR after the upstream investigation changed the diagnosis.

The reported failure was not fixed by disabling compilation: the checkpoint lacked the required MTP weights, leaving the instantiated MTP layers unpopulated and producing NaNs. Separately, current vllm-xpu-kernels has a native grouped-topk kernel, and intel/torch-xpu-ops #4986 merged the NaN-safe `topk` fix.

The proposed eager-router workaround no longer matches the root cause and would add an unnecessary XPU-specific graph break. Thanks for tracing this to the checkpoint and Torch behavior.
